iOS: Geolocation: Allow skipping of permission prompts

rigdern opened the pull request:
This change enables developers to tell the Geolocation module to skip its permissions requests so the app can manage the permissions requests on its own. React Native Android's Geolocation module already requires apps to request permissions on their own.

Currently, the Geolocation module automatically makes permissions requests for you based on what you've specified in your property list file. However, the property list file doesn't always represent what permissions the app wants right now. For example, suppose an app sometimes wants location when "in use" and sometimes wants location "always" depending on what app features the user has engaged with. The Geolocation API will request the "always" location permission even if that's not what the app wants for the current scenario.

This change enables developers to tell the Geolocation module to skip permission requests so that the app can explicitly ask for the appropriate permissions. By default, Geolocation will still ask for permissions so this is not a breaking change.

This change adds a method to Geolocation that is not supported by the web spec for geolocation (`setRNConfiguration`). This method includes `RN` in the name to minimize the chances that the web spec will introduce an API with the same name.

**Test plan (required)**

Verified that Geolocation requests permissions by default and when `skipPermissionRequests` is `false`. Verified the permission requests are skipped when `skipPermissionRequests` is `true`. Also, my team is using this change in our app.

Adam Comella
Microsoft Corp.

chirag04 commented on Libraries/Geolocation/RCTLocationObserver.m:
```diff
@@ -135,7 +149,9 @@ - (dispatch_queue_t)methodQueue
 - (void)beginLocationUpdatesWithDesiredAccuracy:(CLLocationAccuracy)desiredAccuracy distanceFilter:(CLLocationDistance)distanceFilter
 {
-  [self requestAuthorization];
+  if (!_locationConfiguration.skipPermissionRequests) {
```
wondering if we can just check if permissions are already granted instead of a skip flag

rigdern replied:
@chirag04 That's an interesting suggestion. A downside to that design is it'll open the door to a tricky to catch app bug. Suppose the app wants to request permissions on its own but fails to. With the current implementation, the app will see an exception due to not having the proper permissions. With your suggestion, RN will request permissions on behalf of the app making it difficult for the developer to realize there's a bug in their app.

javache commented on Libraries/Geolocation/Geolocation.js:
```diff
@@ -75,6 +79,23 @@ type GeoOptions = {
+    *
+    */
+  setRNConfiguration: function(
+    geo_config: GeoConfiguration
```
Use camelCase. In this case, just name the argument `config`

javache commented on Libraries/Geolocation/RCTLocationObserver.m:
```diff
@@ -160,6 +176,11 @@ - (void)timeout:(NSTimer *)timer
 
 #pragma mark - Public API
 
+RCT_EXPORT_METHOD(setRNConfiguration:(RCTLocationConfiguration)config)
```
Why the RN in the name? Can this be just `setConfiguration:`

javache replied:
Sorry, missed that in your original description. I think it's a good idea to avoid potential conflicts with the web spec. We can name the native module however we want though, so let's call the native method `setConfiguration` and the API on geolocation `setReactConfiguration`?

javache commented on ReactAndroid/src/main/java/com/facebook/react/modules/location/LocationModule.java:
```diff
@@ -104,6 +104,11 @@ private static LocationOptions fromReactMap(ReadableMap map) {
     }
   }
 
+  @ReactMethod
```
Remove this completely and instead check on the JS side if the method exists before calling it.

rigdern commented on Libraries/Geolocation/RCTLocationObserver.m:
```diff
@@ -160,6 +176,11 @@ - (void)timeout:(NSTimer *)timer
 
 #pragma mark - Public API
 
+RCT_EXPORT_METHOD(setRNConfiguration:(RCTLocationConfiguration)config)
```
@javache "This method includes RN in the name to minimize the chances that the web spec will introduce an API with the same name."

hramos commented:
I've taken a look at the current status of this PR, and based on the comments left by other reviewers as well as the conflicts, I'm marking it as "needs-revision".

rigdern commented:
@hramos thanks for the ping.

@javache thanks for reviewing this. I pushed an update which addresses your comments. As described in my other comment, I left the method `setRNConfiguration` with the RN prefix. I justified why I used the prefix but if you'd prefer, I can remove the prefix. Let me know.

rigdern commented:
PR #15067 also solves this problem. We should figure out which PR to take. See my comment (https://github.com/facebook/react-native/pull/15067#issuecomment-325207098) for a comparison of these PRs.

ngandhy commented:
@rigdern - when I add the line "navigator.geolocation.setRNConfiguration({'skipPermissionRequests':true});" to my code the location times out every time. 

The second I remove the line of code, everything works again. Any tips / advice?

ngandhy commented:
@rigdern @javache @hramos - Found the issue. _locationManager is never being set if you skip the permission request using skipPermissionsRequest:true. This is because _locationManager is being set INSIDE the authorization method. 

Adding the following inside beginLocationUpdatesWithDesiredAccuracy fixes the issue.

if (!_locationManager) {
    _locationManager = [CLLocationManager new];
    _locationManager.delegate = self;
  }

rigdern commented:
@ngandhy That looks like a bug. Good find! Do you want to send a PR?

Adam Comella
Microsoft Corp.

ngandhy commented:
kk, I'll give it a try :)

Also, I've seen your features, fixes, and all the other stuff you're doing for the community and just wanted to say THANKS!  I'm sure you guys don't hear it as much as you should

rigdern commented:
> Also, I've seen your features, fixes, and all the other stuff you're doing for the community and just wanted to say THANKS! I'm sure you guys don't hear it as much as you should

@ngandhy, thanks, I'm glad you've found our work to be valuable :)
